### PR TITLE
Make use of REMOVE_NODE return constant in refator() method

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

**Review by samsonasik — CHANGES_REQUESTED** [review]
When refactor `return NodeTraverser::REMOVE_NODE;`, the 

```php
        /** @var non-empty-array<Node>|Node $refactoredNode */
        $this->createdByRuleDecorator->decorate($originalNode, $originalNode, static::class);
```

need to be called to print removed node line and rules on `postRefactorProcess()`


```php
        $originalNode ??= $node;

        // node is removed, nothing to post process
        if (is_int($refactoredNode)) {
  
           /** @var non-empty-array<Node>|Node $refactoredNode */
           $this->createdByRuleDecorator->decorate($refactoredNode, $originalNode, static::class);

            return $originalNode;
        }

```

**Comment by TomasVotruba** [comment]
@samsonasik Thanks a lot, I could not figure what notification code put where :)

File: src/Rector/AbstractRector.php [hunk 1]
```diff
@@ -358,7 +358,14 @@ private function postRefactorProcess(Node|null $originalNode, Node $node, Node|a
     {
         // node is removed, nothing to post process
         if (is_int($refactoredNode)) {
-            return $originalNode ?? $node;
+            $currentNode = $originalNode ?? $node;
+            // mark removed node
+            if ($refactoredNode === NodeTraverser::REMOVE_NODE) {
+                /** @var non-empty-array<Node>|Node $refactoredNode */
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php]: this `@var` line can be removed.

```suggestion
```

File: src/Rector/AbstractRector.php [hunk 2]
```diff
@@ -358,7 +358,14 @@ private function postRefactorProcess(Node|null $originalNode, Node $node, Node|a
...
+                $this->createdByRuleDecorator->decorate($originalNode, $originalNode, static::class);
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php]:
```suggestion
                $this->createdByRuleDecorator->decorate($currentNode, $currentNode, static::class);
```

File: src/Rector/AbstractRector.php [hunk 3]
```diff
@@ -358,7 +358,13 @@ private function postRefactorProcess(Node|null $originalNode, Node $node, Node|a
     {
         // node is removed, nothing to post process
         if (is_int($refactoredNode)) {
-            return $originalNode ?? $node;
+            $currentNode = $originalNode ?? $node;
+            // mark removed node
+            if ($refactoredNode === NodeTraverser::REMOVE_NODE) {
+                $this->createdByRuleDecorator->decorate($currentNode, $currentNode, static::class);
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php]: it seems needs a node instead of original node, I will check.

**Comment by TomasVotruba** [comment]
Step to reproduce locally:

```bash
cd e2e/applied-rule-removed-node && php ../e2eTestRunner.php
cd ../..
```

**Comment by TomasVotruba** [comment]
CI is passing :partying_face:

File: src/Rector/AbstractRector.php [hunk 4]
```diff
@@ -238,6 +238,11 @@ final public function enterNode(Node $node)
         // @see NodeTraverser::* codes, e.g. removal of node of stopping the traversing
         if ($refactoredNode === NodeTraverser::REMOVE_NODE && $originalNode instanceof Node) {
             $this->toBeRemovedNodeHash = spl_object_hash($originalNode);
+
+            // notify this rule changing code
+            $rectorWithLineChange = new RectorWithLineChange(static::class, $originalNode->getLine());
+            $this->file->addRectorClassWithLine($rectorWithLineChange);
```
Comment (samsonasik) [inline src/Rector/AbstractRector.php:244]: The decorator information need to be applied too, I will create separate PR/

**Pull request merged**